### ISSUE TITLE
Spike: explore and document the writing of metadata to the same entity from different sources

### ISSUE THREAD

**Issue by LavMatt** [ministryofjustice/find-moj-data#287]
We need to test out, understand and document datahub's behaviour when writing to existing entities from different sources.

E.g. We may have scheduled datahub ingestions for a given source (glue/athena/dbt), the resulting catalogue items may be supplemented with information that is not contained within the scheduled ingestion. 

### Questions we should look to answer
- How are writes to the same entity from different methods handled by datahub? Will want to look at native ingestions, our python library, and custom ingestion sources.
- Do we need to configure anything specifically within datahub or our existing ingestion methods to handle this so no information is ever lost or overwritten?
- Do different entry methods to datahub need to be handled differently?
- How we ensure catalogue entities are not duplicated from different source ingestions? e.g. dbt and glue/athena datahub ingestions will ingest (i expect) the same entity twice.

These properties look like the best place to start exploring:https://datahubproject.io/docs/generated/metamodel/entities/dataset/#editabledatasetproperties, https://datahubproject.io/docs/generated/metamodel/entities/dataset/#editableschemametadata

This spike should allow us to present a clear recommendation for the most effective flow from AP managed data to Datahub. And it'll cover the case for most of the data contained within the catalogue currently, e.g. databases./tables held in glue/athena supplemented by descriptions held outside of glue/athena 

(saying that we have not yet setup any scheduled ingestions for glue/athena)

**Comment by LavMatt** [comment]
- How are writes to the same entity from different methods handled by datahub? Will want to look at native ingestions, our python library, and custom ingestion sources.
> Datahub uses the urn to know whether it should writes to an existing entity. The default behaivour is to overwrite, and this means that useful information that's been added, such as descriptions are wiped by an automated ingestion using the current version of our python library and its upsert method. I believe i have workarounds to varying levels of satisfaction for the overwrites which will be detailed below.

- Do we need to configure anything specifically within datahub or our existing ingestion methods to handle this so no information is ever lost or overwritten?
> In short yes. 
> We will need to change our library upsert methods to push certain data to different aspects in datahub, and customise our ingestion recipes supplementing them with a python script. This will likely mean we might want to orchestrate all ingestion from outside Datahub.

- Do different entry methods to datahub need to be handled differently?
> TODO

- How we ensure catalogue entities are not duplicated from different source ingestions? e.g. dbt and glue/athena datahub ingestions will ingest (i expect) the same entity twice.
> TODO

**Comment by LavMatt** [comment]
### Glue/Athena DataHub scheduled ingestion supplemented by metadata push using our python library

This covers the scenario for most of the data we currently have in the catalogue. In a production environment, we would have a selection of databases ingested to datahub on a schedule from Glue/Athena in the analytical platform. 

These data often will not have certain useful metadata, such as descriptions, tags and the custom properties we want to collect, so we supplement these either from sources such as the data discovery tool, or experts editing descriptions/tags/custom properties of the data.

We need to ensure the catalogue is aligned with the latest schema as in the database (columns, types) via scheduled ingestions and also that we are not overwriting any of the enriched metadata written to the catalogue from a different source.

I’ve run test ingestions to understand datahub’s behaviour for the above scenario.

### Issues
#### 1 - Database URNs differ
Container (databases) URNs (uniform resource names) are not created by datahub as we create them in our python library. They are created as a guid. This means our python library will create a new container entity rather than update the existing one.

**Solution**: Make our python library generate database urns that match datahub’s urns.
Code below achieves that:
```
from datahub.emitter.mcp_builder import DatabaseKey

DatabaseKey(
        database="some_database",
        platform="athena",
        env="PROD",
        backcompat_env_as_instance=True,
    ).as_urn()
```
#### 2 - Ingests all glue jobs as default
Where not specified in Datahub’s Glue ingestion recipe, datahub will also ingest all glue jobs in the target aws account. These are transform jobs, which we have not really discussed about displaying and not really (imo) of much use to many users looking for data.

**Solution**: In the recipe the key extract_transforms can be set to true or false to either ingest or not ingest glue jobs.

#### 3 - Database owner is set to aws account by default
Where not specified in Datahub’s Glue ingestion recipe, datahub will set the owner of a database to the account number it was ingested from.

**Solution**: set extract_owners to false

#### 4 - Descriptions are overwritten by Datahub Glue ingestion
Our python library used as is to supplement data with descriptions (e.g. from the data discovery tool) has those descriptions overwritten the next time a DataHub glue ingestion is run.

**Solution**: We need to update our python library upsert methods to use the Editable.. Versions of aspects for certain properties. Some example code for a dataset(table):
```
editable_dataset_properties = EditableDatasetPropertiesClass(
    description=metadata.description,
)

metadata_event = MetadataChangeProposalWrapper(
    entityUrn=dataset_urn,
    aspect=editable_dataset_properties,
)
self.graph.emit(metadata_event)


editable_schema_info: list[EditableSchemaFieldInfoClass] = [
    EditableSchemaFieldInfoClass(
        fieldPath=column["name"],
        description=column["description"],
    )
    for column in metadata.column_details
]
editable_schema_properties = EditableSchemaMetadataClass(
    editableSchemaFieldInfo=editable_schema_info
)

metadata_event = MetadataChangeProposalWrapper(
    entityType="dataset",
    changeType=ChangeTypeClass.UPSERT,
    entityUrn=dataset_urn,
    aspect=editable_schema_properties,
)
self.graph.emit(metadata_event)
```

#### 5 - Custom Properties are overwritten
Unlike the above there is no editable version of custom properties. They are defined within DatasetPropertiesClass and EditableDatasetPropertiesClass does not include custom properties. As such any new ingestion to an entity overwrites its custom properties. What we include in custom properties are items like where_to_access_data, dpia_required, sensitivty_level and these could obviously grow as we scale up.

**Solution**: See [datahub slack](https://datahubspace.slack.com/archives/CUMUWQU66/p1687295744732159) for thread about this issue. The proposed work around is to use a transformer in the ingestion recipe https://datahubproject.io/docs/metadata-ingestion/docs/transformer/dataset_transformer/#config-details-15
And
https://datahubproject.io/docs/metadata-ingestion/docs/transformer/dataset_transformer/#writing-a-custom-transformer-from-scratch

In summary, can define a custom transformer class to retain properties that already exist in datahub, where otherwise they’d be overwritten.

This would mean somewhat customising the glue ingestion and would need to be run from outside datahub via the cli, possibly in github actions.

Example of a custom class used by a transformer in a recipe can be seen in the draft PR here https://github.com/ministryofjustice/data-catalogue-metadata/pull/20

**Comment by LavMatt** [comment]
Still to do - Better understand how glue/athena and dbt ingestions will works together

**Comment by LavMatt** [comment]
some more to do:

- See what happens with glue when descriptions exist and we use editable ones too
- Can we hold custom properties (key/value pairs) in the glue catalog

**Comment by MatMoore** [comment]
Some follow up actions to properly handle DBT and any other sources that create sibling entities:

1. When fetching the details page, traverse sibling entities and merge properties (the sibling with isPrimary should take precedence)
2. When fetching the search page, fetch siblings, merge properties, ensure there are no duplicates (refer to [Datahub React Frontend](https://github.com/datahub-project/datahub/blob/9debbdd4a93ba31f107c3e894031532142469174/datahub-web-react/src/app/search/utils/combineSiblingsInSearchResults.ts))

**Comment by MatMoore** [comment]
Summary of DBT findings from https://docs.google.com/document/d/1cLDJQl3Jgyt1RZkKd_V2ViieR232kHVpHxzX5XwMWCA/edit#heading=h.cr74trydkf6z

## TL;DR 

I think generally we want a single source to be authoritative per URN/aspect. Whenever we diverge from that we are creating a ton of complexity. Having separate but linked (sibling) entities is a workaround but I don't think it is a good idea.

## Summary of issues

1. DBT ingestion does not create containers (database) entities for ingested models (tables). Impact: DBT ingestion alone will not create database entities, only dataset ones.
2. DBT ingestion creates two distinct entities with their own URNs: one for the target platform, and one for the DBT platform. It is the frontend's responsibility to combine these for presentation, but we don't do that at the moment (see comment above). These related entities are linked via the SiblingAssociationHook in the consumer job. 
3. DBT seeds create additional *seed* datasets but use a different urn scheme to identically named *table* datasets ingested via the Athena source. These seeds do not pick up schema metadata for some reason.
4. In general, DBT ingestion and Athena ingestion generate different URNs, due to the former missing the catalogue part.

## My conclusions

Mixing and matching ingestion sources for the same metadata seems very error prone. While PATCH behaviour is implemented for owners, tags, and glossary terms, generally ingestions will override metadata. So if we use both DBT and Athena, and the metadata is inconsistent, what we see in the catalogue could change based on which one ran last.

Also, if we [enable stateful ingestion](https://datahubproject.io/docs/metadata-ingestion/docs/dev_guides/stateful/) we may have some unusual edge cases around stale metadata removal that don’t occur in a typical datahub installation.

## Recommendations
We should not set up multiple ingestion sources that ingest the same metadata aspects. So either we use DBT or Athena exclusively, or we configure them in such a way that they don't conflict with each other.

When it comes to manually provided updates to descriptions etc. we should write them to the `EditableSchemaDescription` and `EditableProperties` so that they don't conflict with automatically provided metadata.
> This separates changes made from ingestion pipelines and edits in the UI to avoid accidental overwrites of user-provided data by ingestion pipelines

Tom's suggestion: have a declared hierarchy of source preferences, and create some functionality to check for urn conflicts (likely involving --dry-run) before adding / applying filters to lower priority sources to remove conflicts with higher priority sources

There is an open problem to be solved: what to do with custom properties, since they cannot be written to EditableProperties, therefore they can be overwritten by automated ingestions.

We should disable DBT *seeds* in `entities_enabled`, since
- They are effectively internal to CaDeT
- They duplicate athena tables, which is difficult to reconcile in the catalogue
- They may be useful for lineage, but we don’t have this working yet

We need to do extra work in our frontend to handle sibling relationships. If we want to merge entities we can recreate the logic from Datahub’s frontend, but an MVP solution might be to exclude the DBT platform from our search, and then if we are missing information on the details page we can pull it in via the siblings relationship.




## Options
To avoid complexity we could source AP data primarily from DBT, and exclude from Athena ingestions anything that is ingested via DBT. However this approach has many caveats:
- This would prevent us picking up column descriptions from the glue catalogue
- This would prevent us ingesting other aspects from athena, like profiling information
- We will still need to ingest non-derived tables from Athena if no tables are derived for them in DBT
- If the database containers are populated from Athena, but tables are populated via DBT, then they won’t be connected properly

Alternatively, since DBT doesn’t ingest any containers for databases, maybe we could use the athena source for everything, and then ingest DBT separately. The problem with this is we would need to do something manual to connect up the DBT ingested entities with the Athena ingested entities:
- If we have the same thing in both, we will need to manually add sibling aspects based on some custom code
- We should use a distinct target platform instance from DBT so that each ingestion source writes to its own platform

We could optionally disable sources from the DBT ingestion
- This will allow us to ingest all the source tables via the athena ingestion
- But this will break the lineage of derived tables

**Comment by alex-vonfeldmann** [comment]
i may have missed this / this might be covered in the above, but is there an additional scenario where data / metadata has a major update so that any manually added description (added via datahub) might become outdated / wrong?

**Comment by LavMatt** [comment]
### Outcome
We discussed his spike as a group and centred the discussion around these slides which aim to summarise the above in a more simple manner: https://docs.google.com/presentation/d/19xKko6kq-dot4yD6qukMLK7Xm-9_efzbqWaVNHshjdQ/edit#slide=id.g13cb8a0db7d_0_0

Outcomes from the discussion were:
- We will focus on DBT (create-a-derived-table) ingestion
  - we will not attempt to ingest dbt sourced data from glue/athena in tandeem
  - we will not assign a `target_platform` in the dbt ingestion recipe - this will result in one entity created in datahub per dbt model.
  - we will not allow updates to dbt metadata via any other source, be it datahub ui or our datahub python library
- We will delete what we have currently ingested from Glue/Athena.

We may have to revisit and consider some of the recommendations or suggestions in this spike if issues arise during future development, when onboard different data from different sources (e.g. data discovery tool)
